src/price_data.py

```diff
@@ -414,3 +420,166 @@ def get_cost(
         if isinstance(tr, transaction.SoldCoin):
             return price * tr.sold
         raise NotImplementedError
+
+    def get_candles(self, start: int, stop: int, symbol: str, exchange: str) -> list:
+        exchange_class = getattr(ccxt, exchange)
+        exchange_obj = exchange_class()
+        if exchange_obj.has["fetchOHLCV"]:
+            sleep(exchange_obj.rateLimit / 1000)  # time.sleep wants seconds
+            # get 2min before and after range
+            startval = start - 1000 * 60 * 2
+            rang = max(int((stop - start) / 1000 / 60) + 2, 1)
+            return list(exchange_obj.fetch_ohlcv(symbol, "1m", startval, rang))
+        else:
+            log.error("fetchOHLCV not implemented on exchange, skipping ohlcv")
+            # shouldnt happen technically because exchanges are filterd for fetchohlcv
+            return []
+
+    def _get_bulk_pair_data_path(
+        self,
+        operations: list,
+        coin: str,
+        reference_coin: str,
+        preferredexchange: str = "binance",
+    ) -> list:
+        def merge_prices(a: list, b: list = []) -> list:
+            prices = []
+            if not b:
+                return a
+            for i in a:
+                factor = None
+                for j in b:
+                    if i[0] == j[0]:
+                        factor = j[1]
+                        break
+                prices.append((i[0], i[1] * factor))
+            return prices
+
+        timestamps: list = []
+        timestamppairs: list = []
+        maxminutes = (
+            300  # coinbasepro only allows a max of 300 minutes need a better solution
+        )
+        timestamps = [op.utc_time for op in operations]
+        if not preferredexchange:
+            preferredexchange = "binance"
+
+        current_first = None
+        for timestamp in timestamps:
+            if (
+                current_first
+                and current_first + datetime.timedelta(minutes=maxminutes - 4)
+                > timestamp
+            ):
+                timestamppairs[-1].append(timestamp)
+            else:
+                current_first = timestamp
+                timestamppairs.append([timestamp])
+        datacomb = []
+        for batch in timestamppairs:
+            # ccxt works with timestamps in milliseconds
+            first = misc.to_ms_timestamp(batch[0])
+            last = misc.to_ms_timestamp(batch[-1])
+            firststr = batch[0].strftime("%d-%b-%Y (%H:%M)")
+            laststr = batch[-1].strftime("%d-%b-%Y (%H:%M)")
+            log.info(
+                f"getting data from {str(firststr)} to {str(laststr)} for {str(coin)}"
+            )
+            path = self.path.getpath(
```

Comment:
This currently has a small performance improvement that is possible:
currently for every batch the paths are recomputed because i filter them at the end for the matching path to my criteria.
but the paths as a list stay the same and only the filtering and computing of the active timeframe is done again which both are very simple computations on the second execution, but the paths which stay the same are computed again which is a little bit slow with many pairs.

It works, but this would be a place of improvement